File: backend/src/main/java/com/bondspace/service/MemoryService.java
package com.bondspace.service;

import com.bondspace.domain.model.Memory;

import java.util.List;

public interface MemoryService {

    // Create a new memory
    Memory createMemory(String type, String name, String description, List<String> tags, String textContent, String mediaContent);

    // Retrieve a memory by its ID
    Memory getMemoryById(int memoryId);

    // Retrieve all memories from a space
    List<Memory> getAllMemoriesFromASpace(int spaceId);

    // Add tags to a memory
    boolean addTagsToMemory(int memoryId, List<String> tags);

    // Remove tag from a memory
    boolean removeTagFromMemory(int memoryId, String tag);

    // Add media content to a memory
    boolean addMediaContentToMemory(int memoryId, String mediaContent);

    // Replace media content for a memory
    boolean replaceMediaContentInMemory(int memoryId, String newMediaContent);

    // Remove media content from a memory
    boolean removeMediaContentFromMemory(int memoryId, String mediaContent);

    // Update the details of an existing memory
    boolean updateMemory(String type, String name, String description, String textContent);

    // Delete an existing memory
    boolean deleteMemory(int memoryId);

    // Filter a space's memories by tag
    List<Memory> filterMemoriesByTag(int spaceId, String tag);

    // Filter a space's memories by name
    List<Memory> filterMemoriesByName(int spaceId, String name);

}

File: frontend/angular-bondspace-web-app/src/app/service/memory.service.ts
// memory.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class MemoryService {
  private readonly API_URL = 'http://localhost:8080/api/memories';

  constructor(private http: HttpClient) {}

  createMemory(spaceId: number, name: string, tags: string[], textContent: string): Observable<any> {
    return this.http.post(`${this.API_URL}/create`, {
        spaceId,
        name,
        tags,
        textContent
    }, { withCredentials: true });
  }

  getMemoryById(memoryId: number): Observable<any> {
    return this.http.get(`${this.API_URL}/${memoryId}`, { withCredentials: true });
  }

  deleteMemory(memoryId: number): Observable<any> {
    return this.http.delete(`${this.API_URL}/${memoryId}`, { withCredentials: true });
  }

  getSpaceTags(spaceId: number): Observable<string[]> {
    return this.http.get<string[]>(`${this.API_URL}/space/${spaceId}/tags`, { withCredentials: true });
  }
}